bitcoin/src/address.rs

```diff
@@ -32,7 +32,8 @@ use core::fmt;
 use core::marker::PhantomData;
 use core::str::FromStr;
 
-use bech32;
+use bech32::primitives::iter::{ByteIterExt, Fe32IterExt};
+use bech32::{witness_version, Bech32, Bech32m, WitnessVersion};
```

Comment:
I find it odd that `WitnessVersion` lives in `bech32` now. Can you explain the decision more?

Reply:
This wasn't my decision, but I would guess that @tcharding figured that since we have to support witness versions at least in the form of arbitrary bech32 characters (since the bech32 spec itself treats witness versions specially, decoding the witness version character as an 8-bit byte rather than a 5-bit thing), we might as well give it a dedicated type so we could enforce that only the bech32-valid values are used.

I am of two minds on this ... on the one hand it decreases the generality of the library a bit by undermining other checksums that use a wider notion of "witness version" ... but on the other hand, AFAICT only bech32 and bech32m have a notion of "witness version" at all, and the other checksums used in Bitcoin just decode bech32 strings as strings.

Reply:
Note that any checksum that *did* want to have their own non-bip173-like version of witness versions would be able to do this by creating another iterator adaptor, and then adding some new high-level funcitons to wrap it.

Reply:
Yes, it seems to me that the bips are pretty tightly coupled with the concept of witness version and also the same arguments apply to `Network` so I thought we should have types for them.

> Note that any checksum that did want to have their own non-bip173-like version of witness versions would be able to do this by creating another iterator adaptor, and then adding some new high-level funcitons to wrap it.

Unless I'm missing something I think users wanting to opt out of `WitnessVersion` can use `Parsed::new` and the fist byte can have any meaning they want, right?

> I am of two minds on this ... on the one hand it decreases the generality of the library a bit by undermining other checksums that use a wider notion of "witness version" ... but on the other hand, AFAICT only bech32 and bech32m have a notion of "witness version" at all, and the other checksums used in Bitcoin just decode bech32 strings as strings.

So this is my main concern as well, that is why I added a couple of new methods in `primitives` that allow bypassing the `Network` type (using a string HRP). I was steering this towards making the crate root API functions super bip-173/350 specific and really easy to use for the typical Bitcoin usecase. All other niche usecases have to go to `primitives`. 

We hadn't really talked about the `Network` type @apoelstra, to discuss that was one reason for this draft.

Reply:
I don't understand the comment about `Network` being tightly tied to the BIPs. It's fine to add conversion functions between HRPs and rust-bitcoin `Network`s for use in address computation, but both BIP 173 and BIP 350 have tons of test vectors that use arbitrary HRPs.

>Unless I'm missing something I think users wanting to opt out of WitnessVersion can use Parsed::new and the fist byte can have any meaning they want, right?

Ah, yep, I forgot about the `Parsed` type.

Reply:
So, on a high level, definitely yes.

On a lower level I think it'd be really hard to decide what belongs in there. Do the hash types go in there? How about transactions and blocks (which are what are implied by the "primitives" name, though we could rename it to something more suggestive if we don't intend to include transactions and blocks).

Basically whenever I've tried to come up with such a crate, I start thinking "it would be nice for all the POD types to just site in `bitcoin-constants` or whatever ... and I guess also constants" ... but then I feel like in Bitcoin on some level *everything* is "plain old data" and I get nowhere.

Reply:
Warning, musings: I had the same thought "witness should be in a primitives crate", perhaps any time we have this thought its a hint that we have an abstraction wrong. `WitnessVersion` doesn't really fit in a "bech32" crate but it does fit in a "segwit" crate and maybe even in a "segwith address format" crate. Then the question is "does it make sense for a `Script` type to depend on a "segwit address format" crate, perhaps not. (`Scipt` uses `WitnessVersion`.) What about to depend on a "segwit" crate, perhaps. But what the hell goes in a "segwit" crate?

Reply:
I was thinking about types similar to language primitive types, so mostly newtypes with trivial methods. Or from another POV, no-alloc types. `constants` seems to be a good approximation but I don't like the name `constants` to contain types (which is currently the case with our `constants` module which I'd like to move around)

I was considering a separate crate for `Transaction` and its components already but not sure if it helps anything other than not downloading (and compiling) the whole crate for most projects.

Reply:
lol to calling it bitcoin-core just to troll people.

Reply:
In all seriousness I would be totally fine with a `github.com/rust-bitcoin/core` i.e., a `core` crate inside `rust-bitcoin` repo. The reason is that it perfectly describes what it is, core types. And secondly Core Lightning already made the hilarious decision to muddy usage of the word "core". Also, I've been experiencing lately, non bitcoin developers totally well meaningly but kind of annoyingly, refer to people who work on Bitcoin related software as "core developers". Actually I've never thought this before but Bitcoin Core is a silly name for a piece of software that includes a bunch of non-core stuff.

Reply:
> Oh, I think I know what's the correct rule: types that don't depend on anything else from `bitcoin*` except `internals`. They should operate only on Rust types. So if `bitcoin` is analogous to `std` then `bitcoin-primitives` is analogous to `core` (calling it `core` would be obviously confusing).

I like this. Let's try to use this rule of thumb and see if we can make it work.

`primitives` is fine :). `core` is not :P.

Reply:
`bitcoin-primitives` is taken on crates.io already.

Reply:
`rust-bitcoin-primitives` - primitives types used by the 'rust-bitcoin' ecosystem?

We could take `rust-bitcoin-internals` while we are at it.

Reply:
I'd rather not use a `rust-bitcoin` prefix. It feels a bit too redundant to have `rust` in the name of a Rust crate.

Other suggestions are `bitcoin-basics`, `bitocin-data`, `bitcoin-essential`.

Reply:
I'd almost like `bitcoin-entities` if Clean Architecture was more known - because that's what they are named there.

`-primitive-types` is free but kinda long. `-common` is free, short and understandable I think.

I tried ask ChatGPT and it didn't come up with anything else I like.

I also realized the rule of thumb has a flaw: it'd exclude anything that can be parsed and has `FromStr` because of the associated `Err` type (and making it too lax would just import whole `bitcoin` crate). So I think it should make exception for when the type implements any `std` traits that have associated types then the associated type would be in the crate as well and doesn't count as dependency. Further, the rule of thumb could also include some silly types that are used in a specific submodule of `bitcoin` so I think the rule also needs to require that the type is shared among multiple crates.

Further, I think that maybe we're doing this a bit wrong. `bech32` should be probably project-agnostic - so no dependency on `bitcoin` and whatever functionality we need should be added inside the `address` module (crate). Since presumably if people are parsing addresses they are parsing `Address` and if they are parsing something else they don't need that stuff. Note that this still implies wanting to have a shared crate that will be depended on by address crate, network crate and others.